py/yubikey.py

```diff
@@ -77,275 +82,437 @@ def credential_data_to_dict(credentialData):
+                    touch = True
+                else:
+                    raise
+            return code, touch
 
-    def _read_slot_code(self, slot, digits, timestamp, wait_for_touch):
-        with self._descriptor.open_device(TRANSPORT.OTP) as dev:
-            controller = OtpController(dev.driver)
-            code = controller.calculate(
-                slot, challenge=timestamp, totp=True, digits=int(digits),
-                wait_for_touch=wait_for_touch)
-            valid_from = timestamp - (timestamp % 30)
-            valid_to = valid_from + 30
-            return Code(code, valid_from, valid_to)
+    def otp_calculate_all(
+            self, slot1_digits, slot2_digits, timestamp):
+        valid_from = timestamp - (timestamp % 30)
+        valid_to = valid_from + 30
+        entries = []
 
-    def _slot_name(self, slot):
-        return "YubiKey Slot {}".format(slot).encode('utf-8')
+        if slot1_digits:
+            try:
+                code, touch = self._otp_get_code_or_touch(
+                    1, slot1_digits, timestamp)
+                entries.append({
+                    'credential': cred_to_dict(
+                        Credential("Slot 1".encode(), OATH_TYPE.TOTP, touch)),
+                    'code': code_to_dict(
+                        Code(code, valid_from, valid_to)) if code else None
+                })
+            except YkpersError as e:
+                if e.errno == 4:
+                    pass
+                else:
+                    raise
 
-    def needs_validation(self):
-        try:
-            dev = self._descriptor.open_device(TRANSPORT.CCID)
-            return not self._unlock(OathController(dev.driver))
-        except Exception:
-            return True
+        if slot2_digits:
+            try:
+                code, touch = self._otp_get_code_or_touch(
+                    2, slot2_digits, timestamp)
+                entries.append({
+                    'credential': cred_to_dict(
+                        Credential("Slot 2".encode(), OATH_TYPE.TOTP, touch)),
+                    'code': code_to_dict(
+                        Code(code, valid_from, valid_to)) if code else None
+                })
+            except YkpersError as e:
+                if e.errno == 4:
+                    pass
+                else:
+                    raise
+
+        return success({'entries': entries})
+
+    def otp_calculate(self, slot, digits, credential, timestamp):
+        valid_from = timestamp - (timestamp % 30)
+        valid_to = valid_from + 30
+        with self._open_otp() as otp_controller:
+            code = otp_controller.calculate(
+                slot, challenge=timestamp, totp=True,
+                digits=int(digits), wait_for_touch=True)
+            return success({
+                'credential': credential,
+                'code': code_to_dict(Code(code, valid_from, valid_to))
+            })
```

Comment:
Would it make sense to use `_otp_get_code_or_touch()` here as well (even if we're assuming it should always return `touch=False`)?